### PR TITLE
Ser 206 duka client registration

### PR DESCRIPTION
**Pull request by Bahatiroben**
Duka client registration and storing the collected user data into the data table

[SER-206](https://oneacrefund.atlassian.net/browse/SER-206)
[SER-221]( https://oneacrefund.atlassian.net/browse/SER-221)

test from [here](https://telerivet.com/p/0c6396c9/service/SV7684c662f118f82e/edit)

- dial in the shortcode
- add credit officer Id: **1234**
- choose 1 from the menu (registration)
- You are prompted for a farmer account number which can be any of the following scenarios
1. already existing duka client Account: 
     - Angelo: 27507760
    - Sabin: 27507747
2. a one acre fund account that is not registered to a duka district

3.no account number. (want to register a new client)

scenario 1.
already existing duka client Account: 
     - Angelo: 27507760
    - Sabin: 27507747
- enter the above account number
- you should be prompted to enter a invoice id
- confirm the invoice id
(the phone number, invoice id and account number are recorded into a [DataTable](https://telerivet.com/p/0c6396c9/data/dev_5Fduka_5Fclient_5Fregistration))

scenario 2
 a one acre fund account that is not registered to a duka district
   - Angello: 21270605
   - Sabin: 17735964
- add an account number
- you should get a prompt telling you to add a duka account number or enter "0" to register the current account number to a duka district.
- if the account has an outstanding credit which is the case now, you should see a notice telling you to pay all debt.

scenario 4
once you want to register a new client
- enter 0;
- respond to the coming prompts
- nationalId: any 7 or 8 digits number
- phone number: any 10 digits number
- first and last name: any digits
- invoice number: any digits
after completing all the prompts
- a message should be sent to a farmer telling them that they have registered
- you should see a screen confirming the registration
- a record should be made in the above mentioned data table

### CODE REVIEW

File: duka-client/registration/inputHandlers/accountNumberInputHandler.js [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    withBalance.forEach(function(record){
+        outStandingCredit = outStandingCredit + record.Balance;
+    });
+    return outStandingCredit;
+}
+
+module.exports = {
+    handlerName: handlerName,
+    getHandler: function(lang) {
+        return function(input) {
```
Comment (nyabongo) [inline duka-client/registration/inputHandlers/accountNumberInputHandler.js:38]: Try avoid functions that are more than 5-6  lines long. They can often be broken down into calls to smaller functions that better explain  what is  going on. for instance.
``` javascript
function flipItAndReverse(it){
    var flipped = flip(it);
    return reverse(flipped);
}
```
even without seeing the implementation you can easily see what the aim is.

File: duka-client/registration/inputHandlers/accountNumberInputHandler.test.js [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+jest.mock('../../../shared/rosterApi/getClient');
+jest.mock('../../../shared/rosterApi/registerClient');
+jest.mock('../../../shared/rosterApi/getPhoneNumber');
+
+describe.each(['en-ke', 'sw'])('Farmer\' account number input handler', (lang) => {
+    beforeEach(() => {
+        jest.resetAllMocks();
+        getPhoneNumbers.mockReturnValueOnce([
+            {
```
Comment (nyabongo) [inline duka-client/registration/inputHandlers/accountNumberInputHandler.test.js:16]: If you are going to use the same constant in a "beforeEach" it is better to define it outside the so the 'beforeEach' is smaller and easier to  read

**Review by nyabongo — COMMENTED** [review]
I haven't had time to review it all or test it, but I've left some comments, you don't need to make any changes though, the comments are for future work. To merge please ask both @shejasabin  and @DevFojo to approve.

**Comment by shejasabin** [comment]
> Duka client registration and storing the collected user data into the data table
> 
> [SER-206](https://oneacrefund.atlassian.net/browse/SER-206)
> [SER-221](https://oneacrefund.atlassian.net/browse/SER-221)
> 
> test from [here](https://telerivet.com/p/0c6396c9/service/SV7684c662f118f82e/edit)
> 
> * dial in the shortcode
> * add credit officer Id: **1234**
> * choose 1 from the menu (registration)
> * You are prompted for a farmer account number which can be any of the following scenarios
> 
> 1. already existing duka client Account:
>    
>    * Angelo: 27507760
>    * Sabin: 27507747
> 2. a one acre fund account that is not registered to a duka district
> 
> 3.no account number. (want to register a new client)
> 
> scenario 1.
> already existing duka client Account:
> - Angelo: 27507760
> - Sabin: 27507747
> 
> * enter the above account number
> * you should be prompted to enter a invoice id
> * confirm the invoice id
>   (the phone number, invoice id and account number are recorded into a [DataTable](https://telerivet.com/p/0c6396c9/data/dev_5Fduka_5Fclient_5Fregistration))
> 
> scenario 2
> a one acre fund account that is not registered to a duka district
> 
> * Angello: 21270605
> * Sabin: 17735964
> * add an account number
> * you should get a prompt telling you to add a duka account number or enter "0" to register the current account number to a duka district.
> * if the account has an outstanding credit which is the case now, you should see a notice telling you to pay all debt.
> 
> scenario 4
> once you want to register a new client
> 
> * enter 0;
> * respond to the coming prompts
> * nationalId: any 7 or 8 digits number
> * phone number: any 10 digits number
> * first and last name: any digits
> * invoice number: any digits
>   after completing all the prompts
> * a message should be sent to a farmer telling them that they have registered
> * you should see a screen confirming the registration
> * a record should be made in the above mentioned data table

@Bahatiroben for scenario 2 the account number and phone are not being registered in the table, is it expected behavior?

**Comment by Bahatiroben** [comment]
> > Duka client registration and storing the collected user data into the data table
> > [SER-206](https://oneacrefund.atlassian.net/browse/SER-206)
> > [SER-221](https://oneacrefund.atlassian.net/browse/SER-221)
> > test from [here](https://telerivet.com/p/0c6396c9/service/SV7684c662f118f82e/edit)
> > 
> > * dial in the shortcode
> > * add credit officer Id: **1234**
> > * choose 1 from the menu (registration)
> > * You are prompted for a farmer account number which can be any of the following scenarios
> > 
> > 
> > 1. already existing duka client Account:
> >    
> >    * Angelo: 27507760
> >    * Sabin: 27507747
> > 2. a one acre fund account that is not registered to a duka district
> > 
> > 3.no account number. (want to register a new client)
> > scenario 1.
> > already existing duka client Account:
> > 
> > * Angelo: 27507760
> > * Sabin: 27507747
> > 
> > 
> > * enter the above account number
> > * you should be prompted to enter a invoice id
> > * confirm the invoice id
> >   (the phone number, invoice id and account number are recorded into a [DataTable](https://telerivet.com/p/0c6396c9/data/dev_5Fduka_5Fclient_5Fregistration))
> > 
> > scenario 2
> > a one acre fund account that is not registered to a duka district
> > 
> > * Angello: 21270605
> > * Sabin: 17735964
> > * add an account number
> > * you should get a prompt telling you to add a duka account number or enter "0" to register the current account number to a duka district.
> > * if the account has an outstanding credit which is the case now, you should see a notice telling you to pay all debt.
> > 
> > scenario 4
> > once you want to register a new client
> > 
> > * enter 0;
> > * respond to the coming prompts
> > * nationalId: any 7 or 8 digits number
> > * phone number: any 10 digits number
> > * first and last name: any digits
> > * invoice number: any digits
> >   after completing all the prompts
> > * a message should be sent to a farmer telling them that they have registered
> > * you should see a screen confirming the registration
> > * a record should be made in the above mentioned data table
> 
> @Bahatiroben for scenario 2 the account number and phone are not being registered in the table, is it expected behavior?

hey @shejasabin. I have had missed that. but I have just added it.

**Review by shejasabin — APPROVED** [review]
Looks good. maybe one thing to consider with PM on a later ticket is what to do in case of a duplicate duka national ID registration

**Review by DevFojo — APPROVED** [review]
LGTM

**Pull request merged**